prototyper/prototyper/src/sbi/rfence.rs

```diff
@@ -259,71 +264,235 @@ impl rustsbi::Fence for SbiRFence {
+        }
+        pmu_firmware_counter_increment(firmware_event::HFENCE_VVMA_ASID_SENT);
+
+        let flush_size = match validate_address_range(start_addr, size) {
+            Ok(s) => s,
+            Err(e) => return e,
+        };
+
+        remote_fence_process(
+            RFenceContext {
+                start_addr,
+                size: flush_size,
+                asid,
+                vmid: 0,
+                op: RFenceType::HFenceVvmaAsid,
+            },
+            hart_mask,
+        )
+    }
+
+    fn remote_hfence_vvma(&self, hart_mask: HartMask, start_addr: usize, size: usize) -> SbiRet {
+        if !supports_hypervisor_extension() {
+            return SbiRet::not_supported();
+        }
+        pmu_firmware_counter_increment(firmware_event::HFENCE_VVMA_SENT);
+
+        let flush_size = match validate_address_range(start_addr, size) {
+            Ok(s) => s,
+            Err(e) => return e,
+        };
+
+        remote_fence_process(
+            RFenceContext {
+                start_addr,
+                size: flush_size,
+                asid: 0,
+                vmid: 0,
+                op: RFenceType::HFenceVvma,
+            },
+            hart_mask,
+        )
+    }
 }
 
 /// Handles a single remote fence operation.
 #[inline]
 pub fn rfence_single_handler() {
-    let rfence_context = local_rfence().unwrap().get();
-    if let Some((ctx, id)) = rfence_context {
+    let local_rf = match local_rfence() {
+        Some(lr) => lr,
+        // TODO: Or return an error, depending on expected invariants
+        None => panic!("rfence_single_handler called with no local rfence context"),
+    };
+
+    if let Some((ctx, source_hart_id)) = local_rf.get() {
+        let full_flush = (ctx.start_addr == 0 && ctx.size == 0)
+            || (ctx.size == usize::MAX)
+            || (ctx.size > TLB_FLUSH_LIMIT && ctx.size != usize::MAX);
+
         match ctx.op {
-            // Handle instruction fence
             RFenceType::FenceI => unsafe {
                 pmu_firmware_counter_increment(firmware_event::FENCE_I_RECEIVED);
-                asm!("fence.i");
-                remote_rfence(id).unwrap().sub();
+                asm!("fence.i", options(nostack, nomem));
+                remote_rfence(source_hart_id).unwrap().sub();
             },
-            // Handle virtual memory address fence
             RFenceType::SFenceVma => {
                 pmu_firmware_counter_increment(firmware_event::SFENCE_VMA_RECEIVED);
-                // If the flush size is greater than the maximum limit then simply flush all
-                if (ctx.start_addr == 0 && ctx.size == 0)
-                    || (ctx.size == usize::MAX)
-                    || (ctx.size > TLB_FLUSH_LIMIT)
-                {
-                    unsafe {
-                        asm!("sfence.vma");
-                    }
+                if full_flush {
+                    unsafe { asm!("sfence.vma", options(nostack, nomem)) };
```

Comment:
```suggestion
                    unsafe { asm!("sfence.vma", options(nostack)) };
```
[Rust reference](https://doc.rust-lang.org/stable/reference/inline-assembly.html#r-asm.options.supported-options.nomem) points out: When using `nomem`, The compiler also assumes that the assembly code does not perform any kind of synchronization with other threads, e.g. via fences.
Therefore, `nomem` should not be used here or in the following fence instructions.

Reply:
Thanks for the heads-up! You're right— `nomem` shouldn't be used in subsequent fence instructions to ensure proper synchronization!